Update tag unions portion of tutorial (#4445)
roc-lang/roc: TUTORIAL.md

Similarly, the only way to have a function whose type is `a -> a` is if the function's implementation returns
its argument without modifying it in any way. This is known as [the identity function](https://en.wikipedia.org/wiki/Identity_function).

## Tag Unions

We can also annotate types that include tags:

```coffee
colorFromStr : Str -> [Red, Green, Yellow]
colorFromStr = \string ->
    when string is
        "red" -> Red
        "green" -> Green
        _ -> Yellow
```

You can read the type `[Red, Green, Yellow]` as "a *tag union* of the tags `Red`, `Green`, and `Yellow`."

> Tags are always annotated as being part of a tag union, even if that union only has one tag in it. For example:

Reviewer: I found this a bit confusing to read, since tags don't *have* to be annotated, and "tags" seems to be overloaded in both type and value - what about something like
```suggestion
Tags are always part of a tag union type, even if that union only has one tag in it. For example:
```

```coffee
redTag : [Red]
redTag = Red
```

Tag unions can accumulate additional tags based on how they're used in the program. Consider this `if` expression:

```elm
\str ->
    if Str.isEmpty str then
        "it was empty"
    else
        ["it was not empty"]
```

This will be a type mismatch because the two branches have incompatible types: the first branch

> evaluates to a `Str`, but the `else` branch evaluates to a `List Str`!

Reviewer: Maybe worth moving to this after the correct (Result) example, or further describing why tags are useful in that they avoid mismatches here? To me it's not immediately clear why this example is important to understanding tag accumulation.

Author: I think I can actually just drop this example.

the `Result` type (for operations like `Result.withDefault`, which we saw earlier) is a type
alias for a tag union with `Ok` and `Err` tags that each have one payload:

```haskell
Result ok err : [Ok ok, Err err]
```

We just saw how tag unions get combined when different branches of a conditional return different tags. Another way tag unions can get combined is through pattern matching. For example:

```coffeescript
when color is
    Red -> "red"
    Yellow -> "yellow"
    Green -> "green"
```

Here, Roc's compiler will infer that `color`'s type is `[Red, Yellow, Green]`, because
those are the three possibilities this `when` handles.

Just like how `{}` is the type of an empty record, `[]` is the type of an empty tag union.
There is no way to create an empty tag union at runtime, since creating a tag union requires
making an actual tag, and an empty tag union has no tags in it!

This means if you have a function with the type `[] -> Str`, you can be sure that it will
never execute. It requires an argument that can't be provided! Similarly, if you have a
function with the type `Str -> []`, you can call it, but you can be sure it will not terminate

> normally. The only way to implement a function like that is using [infinite recursion](https://en.wikipedia.org/wiki/Infinite_loop#Infinite_recursion), which will either run indefinitely or else crash with a [stack overflow](https://en.wikipedia.org/wiki/Stack_overflow).

Reviewer: or the `crash` keyword, when it lands!

> **Aside:** Empty tag unions can be useful as type parameters. For example, a function with the type
> `List [] -> Str` can be successfully called, but only if you pass it an empty list. That's because
> an empty list has the type `List *`, which means it can be used wherever any type of `List` is
> needed - even a `List []`!
>
> Taking this a step further, a function which accepts a `Result Str []` only accepts a "Result
> which can never be `Err`" - and you could call that function passing something like `Ok "hello"`

> > with no problem.

Reviewer: Is it worth showing an example of when this might be useful, for example in implementing `List.walk` via `List.walkUntil`? This might seem like neat trivia, but not useful for Roc programming without an example.

Author: Yes! I've updated this in 9156a323eacaff4ec5f4b9b8abcb07d64e1ffe2b - this will actually be a nice segue into a future tutorial section I have in mind on I/O that can fail.

1. The `when color is` now has an extra branch: `_ -> "other"`
2. Since this branch matches any tag, the type annotation for the `color` argument changed from the closed union `[Red, Green]` to the _open union_ `[Red, Green]*`.

Also like with open records, you can name the type variable in an open tag union. For example:

```coffee
stopGoOther : [Red, Green]a -> [Stop, Go]a
stopGoOther = \color ->
    when color is
        Red -> Stop
        Green -> Go
        other -> other
```

You can read this type annotation as "`stopGoOther` takes either a `Red` tag, a `Green` tag, or some other tag. It returns either a `Stop` tag, a `Go` tag, or any one of the tags it received in its argument."

So let's say you called this `stopGoOther` function passing `Foo "hello"`. Then the `a` type variable would be the closed union `[Foo Str]`, and `stopGoOther` would return a union with the type `[Stop, Go][Foo Str]` - which is equivalent to `[Stop, Go, Foo Str]`.

> **Aside:** You can annotate a function as returning an open union with no named type variable -
> such as `Str -> [Red, Green]*` - but since returning doesn't involve pattern matching, this

> > is no different from returning a closed tag union. It's a more verbose way to say the same thing.

Reviewer: This is not totally correct, because there is no such thing as returning a closed tag union anymore. You can only return a tag union whose "open-ness" is inferred.

I would suggest removing this entirely, in part because the compiler will now issue a warning if you have an annotation like `Str -> [Red, Green]*` saying that the `*` is unnecessary. Or, perhaps rephrasing to something like

> Functions are always inferred to return open unions, since the caller of a function can use a return value in any context where a larger open union might be expected.

Author: I forgot about the warning! I'll just delete it then. 👍
